skyportal/models.py

```diff
@@ -103,55 +211,375 @@ def user_to_dict(self):
 User.to_dict = user_to_dict
 
 
-def is_readable_by(self, user_or_token):
-    """Generic ownership logic for any `skyportal` ORM model.
+def _is_accessible(self, user_or_token, access_func):
+    cls = type(self)
+    return (
+        DBSession().query(access_func(user_or_token)).filter(cls.id == self.id).scalar()
+    )
 
-    Models with complicated ownership logic should implement their own method
-    instead of adding too many additional conditions here.
-    """
-    if hasattr(self, 'tokens'):
-        return user_or_token in self.tokens
-    if hasattr(self, 'groups'):
-        return bool(set(self.groups) & set(user_or_token.accessible_groups))
-    if hasattr(self, 'group'):
-        return self.group in user_or_token.accessible_groups
-    if hasattr(self, 'users'):
-        if hasattr(user_or_token, 'created_by'):
-            if user_or_token.created_by in self.users:
-                return True
-        return user_or_token in self.users
-
-    raise NotImplementedError(f"{type(self).__name__} object has no owner")
-
-
-def is_modifiable_by(self, user):
-    """Return a boolean indicating whether an object point can be modified or
-    deleted by a given user.
-
-    Parameters
-    ----------
-    user: `baselayer.app.models.User`
-       The User to check.
 
-    Returns
-    -------
-    readable: bool
-       Whether the Object can be modified by the User.
-    """
+def _bulk_retrieve(
+    cls, user_or_token, opname, accessible_pair_func, required_attrs, options=[]
+):
+    for attr in required_attrs:
+        if not hasattr(cls, attr):
+            raise TypeError(
+                f'{cls} does not have the attribute "{attr}", '
+                f'and thus does not expose the interface that is needed '
+                f'to check if {opname}.'
+            )
 
-    if not hasattr(self, 'owner'):
+    if isinstance(user_or_token, User):
+        accessibility_target = user_or_token.id
+    elif isinstance(user_or_token, Token):
+        accessibility_target = user_or_token.created_by_id
+    else:
         raise TypeError(
-            f'Object {self} does not have an `owner` attribute, '
-            f'and thus does not expose the interface that is needed '
-            f'to check for modification or deletion privileges.'
+            f'Invalid argument passed to user_or_token, '
+            f'expected User or Token, got '
+            f'{user_or_token.__class__.__name__}'
+        )
+
+    pairs = accessible_pair_func(cls.__table__, User.__table__).alias()
+
+    return (
+        DBSession()
+        .query(cls)
+        .join(User, sa.literal(True))
+        .join(pairs, sa.and_(User.id == pairs.c.user_id, cls.id == pairs.c.cls_id))
+        .filter(User.id == accessibility_target)
+        .options(options)
+    )
```

Comment:
@dannygoldstein I'm probably missing something here, but do we really need this pattern of joining the accessible pairs against all possible pairs? Is there a reason we can't just filter directly on the accessible pairs table? Something like `DBSession().query(cls).join(pairs, cls.id == pairs.c.cls_id).filter(pairs.c.user_id == accessibility_target).options(options)`?

Reply:
I just tried this and it seems to return the same results. Only tried it on the one candidates query you and Ari were trying last night for benchmarking though. Very very marginal improvements on the performance as well (like 4-5ms on that query). Could probably do something similar for the `_is_accessible_sql` as well.